apiserver/test/integration/clientset_test.go

```diff
@@ -1349,3 +1355,142 @@ func testIPAMConfigClient(client calicoclient.Interface, name string) error {
+	blockAffinityClient := client.ProjectcalicoV3().BlockAffinities()
+	blockAffinity := &v3.BlockAffinity{
+		ObjectMeta: metav1.ObjectMeta{Name: name},
+
+		Spec: v3.BlockAffinitySpec{
+			CIDR:  "10.0.0.0/24",
+			Node:  "node1",
+			State: "pending",
+		},
+	}
+	libV3BlockAffinity := &libapiv3.BlockAffinity{
+		ObjectMeta: metav1.ObjectMeta{Name: name},
+
+		Spec: libapiv3.BlockAffinitySpec{
+			CIDR:    "10.0.0.0/24",
+			Node:    "node1",
+			State:   "pending",
+			Deleted: "false",
+		},
+	}
+	ctx := context.Background()
+
+	// Calico libv3 client instantiation in order to get around the API create restrictions
+	// TODO: Currently these tests only run on a Kubernetes datastore since profile creation
+	// does not work in etcd. Figure out how to divide this configuration to etcd once that
+	// is fixed.
+	config := apiconfig.NewCalicoAPIConfig()
+	config.Spec = apiconfig.CalicoAPIConfigSpec{
+		DatastoreType: apiconfig.Kubernetes,
+		EtcdConfig: apiconfig.EtcdConfig{
+			EtcdEndpoints: "http://localhost:2379",
+		},
+		KubeConfig: apiconfig.KubeConfig{
+			Kubeconfig: os.Getenv("KUBECONFIG"),
+		},
+	}
+	apiClient, err := libclient.New(*config)
+	if err != nil {
+		return fmt.Errorf("unable to create Calico lib v3 client: %s", err)
+	}
+
+	_, err = blockAffinityClient.Create(ctx, blockAffinity, metav1.CreateOptions{})
+	if err == nil {
+		return fmt.Errorf("should not be able to create block affinity %s ", blockAffinity.Name)
+	}
+
+	// Create the block affinity using the libv3 client.
+	_, err = apiClient.BlockAffinities().Create(ctx, libV3BlockAffinity, options.SetOptions{})
+	if err != nil {
+		return fmt.Errorf("error creating the object through the Calico v3 API '%v' (%v)", libV3BlockAffinity, err)
+	}
+
+	blockAffinityNew, err := blockAffinityClient.Get(ctx, name, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("error getting object %s (%s)", name, err)
+	}
+
+	blockAffinityList, err := blockAffinityClient.List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("error listing BlockAffinity (%s)", err)
+	}
+	if blockAffinityList.Items == nil {
+		return fmt.Errorf("items field should not be set to nil")
+	}
+
+	blockAffinityNew.Spec.State = "confirmed"
+
+	_, err = blockAffinityClient.Update(ctx, blockAffinityNew, metav1.UpdateOptions{})
+	if err == nil {
+		return fmt.Errorf("should not be able to update block affinity %s", blockAffinityNew.Name)
+	}
+
+	err = blockAffinityClient.Delete(ctx, name, metav1.DeleteOptions{})
+	if nil == err {
+		return fmt.Errorf("should not be able to delete block affinity %s", blockAffinity.Name)
+	}
+
+	// Test watch
+	w, err := blockAffinityClient.Watch(ctx, v1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("error watching block affinities (%s)", err)
+	}
+	var events []watch.Event
+	done := sync.WaitGroup{}
+	done.Add(1)
+	timeout := time.After(500 * time.Millisecond)
+	var timeoutErr error
+	// watch for 2 events
```

Comment:
Strictly speaking, we don't need to run this in a goroutine. 

The watch result channel will buffer the events and then we can assert on them synchronously below.